pkg/ccl/changefeedccl/client_bench_test.go

```diff
@@ -0,0 +1,592 @@
+	tb           testing.TB
+	httpServer   http.Server
+	httpListener net.Listener
+
+	tempDir string
+
+	n int
+
+	serverSubprocess struct {
+		sync.Mutex
+		running                    bool // set to true by /serverstart request
+		started, done              chan struct{}
+		sqlAddr, httpAddr, rpcAddr string
+		sslCertsDir                string
+	}
+
+	sink struct {
+		sync.Mutex
+		startTime     time.Time
+		started, done chan struct{}
+		sinkAddr      string
+	}
+	localServer serverutils.TestServerInterface
+	cleanupFunc func()
+	sinkType    sinkType
+}
+
+const (
+	// Arguments for serverStart
+	sqlAddrArg     = "sqladdr"
+	sslCertsDirArg = "sslcertsdir"
+	httpAddrArg    = "httpaddr"
+	listenAddrArg  = "listenaddr"
+
+	// Arguments for sinkStart
+	sinkAddrArg = "sinkaddr"
+
+	serverStartPath = "/serverstart"
+	sinkStartPath   = "/sinkstart"
+	sinkDonePath    = "/sinkdone"
+)
+
+type sinkType int
+
+const (
+	httpSink sinkType = iota
+	sinklessSink
+	kafkaSink
+)
+
+func newChangefeedBenchmark(tb testing.TB, n int, sink sinkType) *changefeedBenchmark {
+	cb := &changefeedBenchmark{tb: tb}
+	var err error
+	cb.tempDir, err = ioutil.TempDir("", "BenchmarkChangeFeed")
+	require.NoError(tb, err)
+	cb.sink.started = make(chan struct{})
+	cb.sink.done = make(chan struct{})
+	cb.serverSubprocess.started = make(chan struct{})
+	cb.sinkType = sink
+	cb.n = n
+	cb.setupMux()
+	return cb
+}
+
+func (cb *changefeedBenchmark) run() {
+	ctx, cancel := context.WithCancel(context.Background())
+	g, ctx := errgroup.WithContext(ctx)
+	defer func() { require.NoError(cb.tb, g.Wait()) }()
+	defer cancel()
+	{
+		// Run the local coordinating test server.
+		l, err := net.Listen("tcp", "127.0.0.1:0")
+		require.NoError(cb.tb, err)
+		cb.httpListener = l
+		g.Go(swallowError(ctx, func() error {
+			return cb.httpServer.Serve(l)
+		}))
+		g.Go(swallowError(ctx, func() error {
+			<-ctx.Done()
+			return cb.httpServer.Close()
+		}))
+	}
+
+	// Run a cockroach server which stores data in a subprocess.
+	// This subprocess will also set up the data set.
+	// It will send a request to the local process when it finishes leading to the
+	// closure of serverSubprocess.started. That server will also set up zone
+	// configurations to have constraints against the the "coordinator" tag that
+	// we'll later user for the server we start inside of this process.
+	g.Go(swallowError(ctx, func() error {
+		return cb.runCockroachServer(ctx)
+	}))
+
+	// Wait for the server subprocess to start up.
+	<-cb.serverSubprocess.started
+
+	// Launch a local SQL server that joins the cluster created in the subprocess.
+	// This server will use the "coordinator" tag and thus avoid hosting any
+	// replicas.
+	cb.runLocalServer()
+
+	// Launch the sink process. The sink process will connect to the local server.
+	// It will create a changefeed and read from the sink until all of the
+	// expected rows have been seen.
+	g.Go(swallowError(ctx, func() error {
+		return cb.runSink(ctx)
```

Comment:
so what is actually being benchmarked here? Is it just how fast we can read all of the rows that workload imported?